File: Day-02/java/cerus/Day02.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Day02 {

    public static void main(final String[] args) throws IOException {
        final List<Game> games = Arrays.stream(input())
                .filter(s -> !s.isEmpty())
                .map(String::trim)
                .map(Day02::parse)
                .toList();

        int sum = games.stream()
                .filter(Game::isPossible)
                .mapToInt(Game::id)
                .sum();
        System.out.println(sum);
        sum = games.stream()
                .mapToInt(Game::calculatePower)
                .sum();
        System.out.println(sum);
    }

    private static Game parse(final String s) {
        String trimmed = s.substring(5);
        final int id = Integer.parseInt(trimmed.substring(0, trimmed.indexOf(':')));
        trimmed = trimmed.substring(trimmed.indexOf(':') + 2);

        final List<Map<Color, Integer>> pairs = new ArrayList<>();
        final String[] pairSplit = trimmed.split("; ");
        for (final String pair : pairSplit) {
            final Map<Color, Integer> pairMap = new HashMap<>();
            final String[] elementSplit = pair.split(", ");
            for (final String element : elementSplit) {
                final String[] split = element.split("\\s+");
                pairMap.put(Color.valueOf(split[1].toUpperCase()), Integer.parseInt(split[0]));
            }
            pairs.add(pairMap);
        }

        return new Game(id, pairs);
    }

    private static String[] input() throws IOException {
        return Files.readString(Path.of("input.txt")).split("\n");
    }

    private enum Color {
        RED(12),
        GREEN(13),
        BLUE(14);

        private final int limit;

        Color(final int limit) {
            this.limit = limit;
        }

        public int getLimit() {
            return this.limit;
        }

    }

    private record Game(int id, List<Map<Color, Integer>> pairs) {

        public int calculatePower() {
            final Map<Color, Integer> fewest = new HashMap<>();
            for (final Map<Color, Integer> pair : this.pairs) {
                for (final Map.Entry<Color, Integer> entry : pair.entrySet()) {
                    fewest.put(entry.getKey(), Math.max(
                            fewest.getOrDefault(entry.getKey(), Integer.MIN_VALUE),
                            entry.getValue()
                    ));
                }
            }
            return fewest.values().stream()
                    .reduce((l, r) -> l * r)
                    .orElse(0);
        }

        public boolean isPossible() {
            return this.pairs.stream()
                    .allMatch(m -> m.entrySet().stream()
                            .allMatch(e -> e.getValue() <= e.getKey().getLimit()));
        }

    }

}

File: Day-02/elixir/icrayix/day02.exs
games = String.split(File.read!("input.txt"), "\n", trim: true)

limits = %{"red" => 12, "green" => 13, "blue" => 14}

defmodule Day02 do
  def max_colors(bag) do
    Enum.map(bag, &String.split(&1, ",", trim: true))
    |> List.flatten()
    |> Enum.map(&String.split/1)
    |> Enum.reduce(Map.new(), fn [amount, color], acc ->
      if Map.get(acc, color, 0) < String.to_integer(amount) do
        Map.put(acc, color, String.to_integer(amount))
      else
        acc
      end
    end)
  end
end

# part 1
Enum.map(games, fn game ->
  ["Game " <> id | sets] = String.split(game, [":", ";"])

  max = Day02.max_colors(sets)

  if Enum.map(limits, fn {color, limit} -> Map.get(max, color) <= limit end)
     |> Enum.all?() do
    String.to_integer(id)
  else
    0
  end
end)
|> Enum.sum()
|> IO.puts()

# part 2
Enum.map(games, fn game ->
  ["Game " <> _id | sets] = String.split(game, [":", ";"])

  Day02.max_colors(sets) |> Enum.reduce(1, fn {_, amount}, acc -> acc * amount end)
end)
|> Enum.sum()
|> IO.puts()
